enterprise/backend/test/metabase_enterprise/advanced_permissions/query_processor/middleware/permissions_test.clj

```diff
@@ -0,0 +1,137 @@
+      (ee.qp.perms/apply-download-limit)
+      (get-in [:query :limit])))
+
+(deftest apply-download-limit-test
+  (let [limited-download-max-rows @#'ee.qp.perms/max-rows-in-limited-downloads]
+    (with-download-perms-for-db (mt/id) :limited
+      (mt/with-current-user (mt/user->id :rasta)
+        (testing "A limit is added to MBQL queries if the user has limited download permissions for the DB"
+          (is (= limited-download-max-rows
+                 (download-limit (mbql-download-query)))))
+
+        (testing "If the query already has a limit lower than the download limit, the limit is not changed"
+          (is (= (dec limited-download-max-rows)
+                 (download-limit (assoc-in (mbql-download-query)
+                                           [:query :limit]
+                                           (dec limited-download-max-rows))))))
+
+        (testing "Native queries are unmodified"
+          (is (= (native-download-query) (ee.qp.perms/apply-download-limit (native-download-query)))))))
+
+    (with-download-perms (mt/id) {:schemas {"PUBLIC" {(mt/id 'venues) :limited
+                                                      (mt/id 'checkins) :full}}}
+      (mt/with-current-user (mt/user->id :rasta)
+        (testing "A limit is added to MBQL queries if the user has limited download permissions for a table which
+                 the query references"
+          (is (= limited-download-max-rows
+                 (download-limit (mbql-download-query)))))
+
+        (testing "If the query does not reference the table, a limit is not added"
+          (is (nil? (download-limit (mbql-download-query 'checkins)))))))))
+
+;; Inspired by the similar middleware wrapper [[metabase.query-processor.middleware.limit-test/limit]]
+(defn- limit-download-result-rows [query]
+  (let [rff (ee.qp.perms/limit-download-result-rows query context.default/default-rff)
+        rf  (rff {})]
+    (transduce identity rf (repeat (inc @#'ee.qp.perms/max-rows-in-limited-downloads) [:ok]))))
+
+(deftest limit-download-result-rows-test
+  (let [limited-download-max-rows @#'ee.qp.perms/max-rows-in-limited-downloads]
+    (with-download-perms-for-db (mt/id) :limited
+      (mt/with-current-user (mt/user->id :rasta)
+        (testing "The number of rows in a native query result is limited if the user has limited download permissions"
+          (is (= limited-download-max-rows
+                 (-> (native-download-query) limit-download-result-rows mt/rows count))))))
+
+    (with-download-perms-for-db (mt/id) :full
+      (mt/with-current-user (mt/user->id :rasta)
+        (testing "The number of rows in a native query result is not limited if the user has full download permissions"
+          (is (= (inc limited-download-max-rows)
+                 (-> (native-download-query) limit-download-result-rows mt/rows count))))))))
+
+(defn- check-download-permisions [query]
+  (:pre (mt/test-qp-middleware ee.qp.perms/check-download-permissions query)))
```

Comment:
`test-qp-middleware` is marked as deprecated but the tests for all the other middleware I looked at that have the `[query rff context]` signature use it, so I figured this is one of the remaining valid use cases. let me know if there's a better approach